File: app/android/app/src/main/java/com/jun/weather/viewmodel/MidWeatherViewModel.java
package com.jun.weather.viewmodel;

import android.app.Application;
import android.util.SparseArray;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;

import com.jun.weather.repository.AppRepository;
import com.jun.weather.repository.web.entity.MidLandItem;
import com.jun.weather.repository.web.entity.MidTempItem;
import com.jun.weather.repository.web.entity.RestResponse;
import com.jun.weather.repository.web.enums.Enums;
import com.jun.weather.util.CLogger;
import com.jun.weather.util.CommonUtils;
import com.jun.weather.viewmodel.entity.FailRestResponse;
import com.jun.weather.viewmodel.entity.MidForecastModel;

import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;

import static com.jun.weather.util.CommonUtils.getDateString;

public class MidWeatherViewModel extends CustomViewModel<SparseArray<MidForecastModel>> {
    public MidWeatherViewModel(@NonNull Application application, AppRepository repository) {
        super(application, repository);

        LiveData<RestResponse<MidLandItem>> data1 = repository.getMidLandData();
        LiveData<RestResponse<MidTempItem>> data2 = repository.getMidTempData();

        data1.observeForever(data -> updateMidForecastModel(data, data2.getValue()));

        data2.observeForever(data -> updateMidForecastModel(data1.getValue(), data));
    }

    private void updateMidForecastModel(RestResponse<MidLandItem> data1, RestResponse<MidTempItem> data2) {
        if(data1 == null || data2 == null) {
            return;
        }

        if(data1.getCode() != Enums.ResponseCode.OK.getValue()) {
            appExecutor.runInUIIO(() -> updateFailData.setValue(
                    new FailRestResponse(data1.getCode(), "land forecast fail,"+data1.getFailMsg()))
            );
        } else if(data2.getCode() != Enums.ResponseCode.OK.getValue()) {
            appExecutor.runInUIIO(() -> updateFailData.setValue(
                    new FailRestResponse(data2.getCode(), "land forecast fail,"+data2.getFailMsg()))
            );
        } else if(data1.getListBody().size() == 0 || data2.getListBody().size() == 0) {
            appExecutor.runInUIIO(() -> updateFailData.setValue(
                    new FailRestResponse(Enums.ResponseCode.EXCEPTION_ERROR.getValue(), "data is empty")
            ));
        } else {
            if(baseDate == null) {
                setBaseDate();
            }

            SparseArray<MidForecastModel> sparseArray = new SparseArray<>();
            parseData(sparseArray, data1, data2);

            appExecutor.runInUIIO(() -> data.setValue(sparseArray));
        }
    }

    private void parseData(SparseArray<MidForecastModel> sparseArray, RestResponse<MidLandItem> data1,
                           RestResponse<MidTempItem> data2) {
        MidLandItem midLandItem = data1.getListBody().get(0);
        MidTempItem midTempItem = data2.getListBody().get(0);

        int keyIdx = isYesterdayBase?3:4;
        for(int i = 3; i <= 10; i++) {
            if(i < 8) {
                sparseArray.put(keyIdx++, getUniModel(i, midLandItem, midTempItem));
            } else {
                sparseArray.put(keyIdx++, getMonoModel(i, midLandItem, midTempItem));
            }
        }
    }

    private MidForecastModel getUniModel(int i, MidLandItem midLandItem, MidTempItem midTempItem) {
        try {
            Object amRainPercentage = midLandItem.getClass().getDeclaredField("rnSt"+i+"Am").get(midLandItem);
            Object pmRainPercentage = midLandItem.getClass().getDeclaredField("rnSt"+i+"Pm").get(midLandItem);
            Object amSky = midLandItem.getClass().getDeclaredField("wf"+i+"Am").get(midLandItem);
            Object pmSky = midLandItem.getClass().getDeclaredField("wf"+i+"Pm").get(midLandItem);

            Object highTemp = midTempItem.getClass().getDeclaredField("taMax"+i).get(midTempItem);
            Object highTempMax = midTempItem.getClass().getDeclaredField("taMax"+i+"High").get(midTempItem);
            Object highTempMin = midTempItem.getClass().getDeclaredField("taMax"+i+"Low").get(midTempItem);
            Object lowTemp = midTempItem.getClass().getDeclaredField("taMin"+i).get(midTempItem);
            Object lowTempMax = midTempItem.getClass().getDeclaredField("taMin"+i+"High").get(midTempItem);
            Object lowTempMin = midTempItem.getClass().getDeclaredField("taMin"+i+"Low").get(midTempItem);

            return new MidForecastModel.Builder()
                    .amRainPercentage(amRainPercentage==null?"":amRainPercentage.toString())
                    .pmRainPercentage(pmRainPercentage==null?"":pmRainPercentage.toString())
                    .amSky(CommonUtils.getSkyIcon(amSky==null?"":amSky.toString()))
                    .pmSky(CommonUtils.getSkyIcon(pmSky==null?"":pmSky.toString()))
                    .highTemp(getTempVal(highTemp==null?"":highTemp.toString(),
                            highTempMax==null?"":highTempMax.toString(),
                            highTempMin==null?"":highTempMin.toString()))
                    .lowTemp(getTempVal(lowTemp==null?"":lowTemp.toString(),
                            lowTempMax==null?"":lowTempMax.toString(),
                            lowTempMin==null?"":lowTempMin.toString()))
                    .baseDate(getDateString(i, baseDate))
                    .build();
        } catch (Exception e) {
            e.printStackTrace();
            return new MidForecastModel.Builder().build();
        }
    }

    private MidForecastModel getMonoModel(int i, MidLandItem midLandItem, MidTempItem midTempItem) {
        try {
            Object rainPercentage = midLandItem.getClass().getDeclaredField("rnSt"+i).get(midLandItem);
            Object sky = midLandItem.getClass().getDeclaredField("wf"+i).get(midLandItem);

            Object highTemp = midTempItem.getClass().getDeclaredField("taMax"+i).get(midTempItem);
            Object highTempMax = midTempItem.getClass().getDeclaredField("taMax"+i+"High").get(midTempItem);
            Object highTempMin = midTempItem.getClass().getDeclaredField("taMax"+i+"Low").get(midTempItem);
            Object lowTemp = midTempItem.getClass().getDeclaredField("taMin"+i).get(midTempItem);
            Object lowTempMax = midTempItem.getClass().getDeclaredField("taMin"+i+"High").get(midTempItem);
            Object lowTempMin = midTempItem.getClass().getDeclaredField("taMin"+i+"Low").get(midTempItem);

            return new MidForecastModel.Builder()
                    .dayRainPercentage(rainPercentage==null?"":rainPercentage.toString())
                    .daySky(CommonUtils.getSkyIcon(sky==null?"":sky.toString()))
                    .highTemp(getTempVal(highTemp==null?"":highTemp.toString(),
                            highTempMax==null?"":highTempMax.toString(),
                            highTempMin==null?"":highTempMin.toString()))
                    .lowTemp(getTempVal(lowTemp==null?"":lowTemp.toString(),
                            lowTempMax==null?"":lowTempMax.toString(),
                            lowTempMin==null?"":lowTempMin.toString()))
                    .baseDate(getDateString(i, baseDate))
                    .build();
        } catch (Exception e) {
            e.printStackTrace();
            return new MidForecastModel.Builder().build();
        }
    }

    private String getTempVal(String val, String high, String low) {
        try {
            return ""+(Integer.parseInt(val)+((Integer.parseInt(high)-Integer.parseInt(low))/2));
        } catch (Exception e) {
            e.printStackTrace();
        }
        return "";
    }

    private boolean isYesterdayBase = false;
    private String baseDate;
    public void updateMidWeather(String landRegionId, String tempRegionId) {
        setBaseDate();

        appExecutor.runInRepositoryIO(() -> {
            repository.updateMidLandData(baseDate, landRegionId);
            repository.updateMidTempData(baseDate, tempRegionId);
        });
    }

    private void setBaseDate() {
        DateTime dateTime = DateTime.now().toDateTime(DateTimeZone.forID("Asia/Seoul"));
        if(dateTime.getHourOfDay() < 6) {
            dateTime = dateTime.minusHours(6+dateTime.getHourOfDay());
            isYesterdayBase = true;
        } else if(dateTime.getHourOfDay() < 18) {
            dateTime = dateTime.minusHours(dateTime.getHourOfDay()-6);
        } else {
            dateTime = dateTime.minusHours(dateTime.getHourOfDay()-18);
        }
        dateTime = dateTime.minusMinutes(dateTime.getMinuteOfHour());
        baseDate = dateTime.toString("yyyyMMddHHmm");
        CLogger.d(baseDate);
    }
}

File: app/android/app/src/main/java/com/jun/weather/viewmodel/MidWeatherViewModel.kt
package com.jun.weather.viewmodel

import android.util.SparseArray
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.jun.weather.repository.AppRepository
import com.jun.weather.repository.web.entity.MidLandItem
import com.jun.weather.repository.web.entity.MidTempItem
import com.jun.weather.repository.web.entity.RestResponse
import com.jun.weather.repository.web.enums.Enums
import com.jun.weather.ui.entity.FailRestResponse
import com.jun.weather.ui.entity.MidForecastModel
import com.jun.weather.util.CLogger
import com.jun.weather.util.CommonUtils
import kotlinx.coroutines.launch
import org.joda.time.DateTime
import org.joda.time.DateTimeZone
import kotlin.reflect.full.memberProperties

class MidWeatherViewModel(repository: AppRepository) : BaseViewModel(repository) {
    private val _data = MutableLiveData<SparseArray<MidForecastModel>>()
    val data: LiveData<SparseArray<MidForecastModel>> = _data

    private fun updateMidForecastModel(data1: RestResponse<MidLandItem>?, data2: RestResponse<MidTempItem>?) {
        if (data1 == null || data2 == null) {
            return
        }
        if (data1.code != Enums.ResponseCode.OK.value) {
            updateFailData.postValue(
                    FailRestResponse(data1.code, "land forecast fail," + data1.failMsg))
        } else if (data2.code != Enums.ResponseCode.OK.value) {
            updateFailData.postValue(
                    FailRestResponse(data2.code, "land forecast fail," + data2.failMsg))
        } else if (data1.listBody!!.isEmpty() || data2.listBody!!.size == 0) {
            updateFailData.postValue(
                    FailRestResponse(Enums.ResponseCode.EXCEPTION_ERROR.value, "data is empty")
            )
        } else {
            if (baseDate == null) {
                setBaseDate()
            }
            val sparseArray = SparseArray<MidForecastModel>()
            parseData(sparseArray, data1, data2)
            _data.postValue(sparseArray)
        }
    }

    private fun parseData(sparseArray: SparseArray<MidForecastModel>, data1: RestResponse<MidLandItem>,
                          data2: RestResponse<MidTempItem>) {
        val midLandItem = data1.listBody!![0]
        val midTempItem = data2.listBody!![0]
        var keyIdx = if (isYesterdayBase) 3 else 4
        for (i in 3..10) {
            if (i < 8) {
                sparseArray.put(keyIdx++, getUniModel(i, midLandItem, midTempItem))
            } else {
                sparseArray.put(keyIdx++, getMonoModel(i, midLandItem, midTempItem))
            }
        }
    }

    private fun getUniModel(i: Int, midLandItem: MidLandItem, midTempItem: MidTempItem): MidForecastModel {
        return try {
            var amRainPercentage: String? = null
            var pmRainPercentage: String? = null
            var amSky: String? = null
            var pmSky: String? = null
            var highTemp: String? = null
            var highTempMax: String? = null
            var highTempMin: String? = null
            var lowTemp: String? = null
            var lowTempMax: String? = null
            var lowTempMin: String? = null

            MidLandItem::class.memberProperties.forEach{
                when(it.name) {
                    "rnSt" + i + "Am" -> amRainPercentage = it.get(midLandItem).toString()
                    "rnSt" + i + "Pm" -> pmRainPercentage = it.get(midLandItem).toString()
                    "wf" + i + "Am" -> amSky = it.get(midLandItem).toString()
                    "wf" + i + "Pm" -> pmSky = it.get(midLandItem).toString()
                }
            }

            MidTempItem::class.memberProperties.forEach {
                when(it.name) {
                    "taMax$i" -> highTemp = it.get(midTempItem).toString()
                    "taMax" + i + "High" -> highTempMax = it.get(midTempItem).toString()
                    "taMax" + i + "Low" -> highTempMin = it.get(midTempItem).toString()
                    "taMin$i" -> lowTemp = it.get(midTempItem).toString()
                    "taMin" + i + "High" -> lowTempMax = it.get(midTempItem).toString()
                    "taMin" + i + "Low" -> lowTempMin = it.get(midTempItem).toString()
                }
            }

            MidForecastModel(
                    lowTemp = getTempVal((lowTemp?.toInt() ?: 0), (lowTempMax?.toInt() ?: 0), (lowTempMin?.toInt() ?: 0)),
                    highTemp = getTempVal((highTemp?.toInt() ?: 0), (highTempMax?.toInt() ?: 0), (highTempMin?.toInt() ?: 0)),
                    amRainPercentage = amRainPercentage ?: "",
                    pmRainPercentage = pmRainPercentage ?: "",
                    amSky = CommonUtils.getSkyIcon(amSky ?: ""),
                    pmSky = CommonUtils.getSkyIcon(pmSky ?: ""),
                    baseDate = CommonUtils.getDateString(i, baseDate)
            )
        } catch (e: Exception) {
            e.printStackTrace()
            MidForecastModel()
        }
    }

    private fun getMonoModel(i: Int, midLandItem: MidLandItem, midTempItem: MidTempItem): MidForecastModel {
        return try {
            var rainPercentage: String? = null
            var sky: String? = null
            var highTemp: String? = null
            var highTempMax: String? = null
            var highTempMin: String? = null
            var lowTemp: String? = null
            var lowTempMax: String? = null
            var lowTempMin: String? = null

            MidLandItem::class.memberProperties.forEach{
                when(it.name) {
                    "rnSt$i" -> rainPercentage = it.get(midLandItem).toString()
                    "wf$i" -> sky = it.get(midLandItem).toString()
                }
            }

            MidTempItem::class.memberProperties.forEach {
                when(it.name) {
                    "taMax$i" -> highTemp = it.get(midTempItem).toString()
                    "taMax" + i + "High" -> highTempMax = it.get(midTempItem).toString()
                    "taMax" + i + "Low" -> highTempMin = it.get(midTempItem).toString()
                    "taMin$i" -> lowTemp = it.get(midTempItem).toString()
                    "taMin" + i + "High" -> lowTempMax = it.get(midTempItem).toString()
                    "taMin" + i + "Low" -> lowTempMin = it.get(midTempItem).toString()
                }
            }

            MidForecastModel(
                    lowTemp = getTempVal((lowTemp?.toInt() ?: 0), (lowTempMax?.toInt() ?: 0), (lowTempMin?.toInt() ?: 0)),
                    highTemp = getTempVal((highTemp?.toInt() ?: 0), (highTempMax?.toInt() ?: 0), (highTempMin?.toInt() ?: 0)),
                    dayRainPercentage = rainPercentage ?: "",
                    daySky = CommonUtils.getSkyIcon(sky ?: ""),
                    baseDate = CommonUtils.getDateString(i, baseDate)
            )
        } catch (e: Exception) {
            e.printStackTrace()
            MidForecastModel()
        }
    }

    private fun getTempVal(`val`: Int, high: Int, low: Int): String {
        try {
            return "" + (`val` + (high - low) / 2)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return ""
    }

    private var isYesterdayBase = false
    private var baseDate: String? = null
    fun updateMidWeather(landRegionId: String?, tempRegionId: String?) {
        setBaseDate()
        viewModelScope.launch {
            updateMidForecastModel(
                    repository.getMidLandData(baseDate, landRegionId),
                    repository.getMidTempData(baseDate, tempRegionId)
            )
        }
    }

    private fun setBaseDate() {
        //중기예보는 6시, 18시에 업데이트됨
        //단기예보는 17시10이후부터 +3일까지 데이터 제공. 이전은 +2일까지 제공
        //중기는 6시의 경우 +3, 18시의 경우 +4 ~ 10일
        var dateTime = DateTime.now().toDateTime(DateTimeZone.forID("Asia/Seoul"))
        if (dateTime.hourOfDay < 6) {
            //전날 18시 예보 요청
            dateTime = dateTime.minusDays(1).withHourOfDay(18)
            isYesterdayBase = true
        } else if (dateTime.hourOfDay < 18) {
            //당일 6시 예보 요청
            dateTime = dateTime.withHourOfDay(6)
        } else {
            //당일 18시 예보 요청
            dateTime = dateTime.withHourOfDay(18)
        }
        dateTime = dateTime.withMinuteOfHour(0)
        baseDate = dateTime.toString("yyyyMMddHHmm")
        CLogger.d(baseDate!!)
    }
}